### PR TITLE
Offer sign-in where the backend refuses a guest, instead of hiding the surface

### PR DESCRIPTION
**Pull request by chelojimenez**
Companion to MCPJam/mcpjam-backend#1451, which refuses anonymous actors at five platform-paid doors (swarm generation, swarm cluster rebuild, wave insights, server quality, run insights) with one shared `sign_in_required` code. Should land with or before that deploy.

## The failure this replaces

Convex prefixes `Server Error` onto every thrown mutation error, and both insight classifiers test for that string before anything else. A guest refusal therefore fell through to the generic branch and set `unavailable` — and `SuiteInsightsCollapsible` returns `null` on `unavailable`. A trial user who had just run their first suite saw **no insights band at all**, and nothing telling them why. The one reader for whom the fix is a single click was the one reader who could not learn that.

## What changed

- **`lib/sign-in-required.ts`** — the client half of the contract. Reads the refusal's own copy from structured `ConvexError.data`, falling back to the serialized payload, because `err.data` does not survive every path a rejection takes to a hook. The **code** is what is matched; the message is only ever read out of the payload that carried it, never guessed from prose.
- **Both classifiers** match it ahead of the generic branch and report `signInRequired` rather than `unavailable`. Not `permanent` either: who is asking can change within a session, what is deployed cannot.
- **`hooks/use-run-insights.ts`** also sets its existing `authRefused` latch, for the reason the workspace-permission branch next to it does: `canRequest` alone does not stop the auto-request for an anonymous hosted visitor, so a guest browsing cohorts would fire one doomed mutation per cohort. An explicit press still clears it. The flag is cleared alongside `error` on a cohort change, so the next cohort never draws a sign-in control under a blank explanation.
- **The run-insights banner** offers *Sign in* where it would have offered *Retry*. Retrying is the one thing that cannot work when the refusal is about who is asking. Same wiring as every other sign-in control — `useAuth().signIn`, tracked `login_button_clicked`, captured return path.
- **Swarm generation.** The proxy forwards the backend's own code in `details.upstreamCode`: the envelope's top-level `code` is the HTTP-shaped one, `FORBIDDEN` for every 403 whatever caused it, and "not a member of this project" is a 403 that signing in does not fix. `SwarmGenerateError` carries `signInRequired`, and the create flow renders `GuestSignInMessage` instead of an error card. The upstream code stays shape-gated by `upstreamErrorCode`, so a WAF interstitial cannot land arbitrary text in a client-read field.

**No `isGuest` branches anywhere.** The backend's answer is the signal, and the auto-request effect already fires once per run (`hasAutoAttemptedRef`), so one refusal per view is the whole cost.

## Tests

New/extended, each with the negative half:

- `lib/__tests__/sign-in-required.test.ts` — structured data, stringified payload, and the four refusals that must **not** be mistaken for this one (`rate_limited`, `FEATURE_UNAVAILABLE`, workspace permissions, missing function).
- `hooks/__tests__/use-run-insights-auth.test.tsx` — the refusal arrives with the `Server Error` prefix on purpose; the band stays visible, navigation does not re-fire, an explicit press does.
- `components/evals/__tests__/use-insight.test.tsx` — and the negative: an undeployed backend stays `unavailable`.
- `components/evals/__tests__/suite-insights-collapsible.test.tsx` — Sign in is offered, Retry is not, and the press does not call the mutation.
- `server/routes/web/__tests__/swarm-generate.test.ts` — the code is forwarded, and none is invented when the backend sent none.

`typecheck:client` clean. Touched suites green (`vitest run` over the evals/hooks/lib/server-route sets: 3460 passing). Note that the server-side suites need `npm run bundle:all` first in a fresh checkout — unrelated to this change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VH8CUKXBpaqrdYUA8RsJSo

---
_Generated by [Claude Code](https://claude.ai/code/session_01VH8CUKXBpaqrdYUA8RsJSo)_

---

> [!NOTE]
> **Medium Risk**
> Touches shared insight error classification and auto-request latching across eval and swarm flows; misclassification could still hide bands or spam mutations, but behavior is heavily test-covered and gated on explicit backend refusal codes.
> 
> **Overview**
> Guest refusals from platform-paid features no longer disappear as “unavailable.” The PR adds **`lib/sign-in-required.ts`** to recognize the backend’s `sign_in_required` / `SIGN_IN_REQUIRED` code (including through Convex’s `Server Error` stringified payloads) and surfaces **`signInRequired`** from **`useInsight`** and **`useRunInsights`** instead of latching **`unavailable`**.
> 
> Insight UIs **stay visible** with the backend message: the run-insights banner shows **Sign in** (not **Retry**), with the usual auth return-path wiring. Hooks **stop repeat doomed auto-requests** for anonymous viewers across run/cohort navigation, while explicit retries and stale in-flight rejections are handled so sign-in mid-session still works.
> 
> Swarm persona generation maps proxy **`details.code`** into **`SwarmGenerateError.signInRequired`** and renders **`GuestSignInMessage`** on Describe instead of duplicating an **`ErrorCard`**. Tests cover classification edge cases, hook latch behavior, UI affordances, and swarm generate 403 forwarding.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 75e8cb14fdc2a8be51cdf865e0420401f6fa731e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Offers Sign in wherever the backend refuses an anonymous guest, instead of hiding the surface or offering a retry that cannot work. Guest refusals used to be classified as `unavailable` and render nothing; now insight bands stay visible with the backend's own message and a Sign in control.

- `lib/sign-in-required.ts` matches the backend's `sign_in_required` code in structured `ConvexError.data` or the serialized payload, never matching prose that merely mentions the code.
- `useInsight` and `useRunInsights` report `signInRequired` ahead of the generic `Server Error` branch and latch it across runs/cohorts, so a guest fires one doomed auto-request instead of one per view.
- An explicit press clears the latch, and a refusal only marks the run it was asked about, so a late rejection never overwrites a newer run or re-latches after sign-in.
- The run-insights banner replaces Retry with Sign in; swarm generation renders `GuestSignInMessage` instead of an `ErrorCard`.
- The swarm proxy forwards the backend's own code in `details.code`, keeping "sign in" distinct from other 403s like "not a member of this project".

Requires the backend's `sign_in_required` refusal to deploy with it.

<sup>Written for commit 8b64ee75380b9bdaa92b1e9f8cf53a163f941206. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/MCPJam/inspector/pull/5235?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/aeb76172-2a1f-4ec5-b668-4e93fac06654?refs=75e8cb14fdc2a8be51cdf865e0420401f6fa731e&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by chelojimenez** [comment]
CI was red on `254e9c8` with four failed checks. **None of them were this PR's** — every one reproduces on unmodified `origin/main` in a scratch worktree. CI lints and typechecks the **merge ref**, so `main`'s tree fails here. This branch was eleven commits behind; `main` is now merged in (conflict-free) and I have ported what could be ported.

### Ported into this PR

All the same shape: a `main` refactor added a hook dependency and left a partial `vi.mock` behind, which **throws on access** rather than returning undefined — taking down a whole pane and failing assertions about unrelated things.

| Suite | Missing mock export |
|---|---|
| `SwarmsTab.perClientEnvLaunch` | `useHost` on `@/hooks/useClients` |
| `SwarmsTab.sessions` | `useConvexAuth` on `convex/react` |
| `SwarmsTab.createFlow` | `useHost` on `@/hooks/useClients` |

That clears **14 of main's 15** failures across those suites. I also lowered the em-dash ratchet entry for `new-swarm-running-step.tsx` from 3 to 2, which is precisely what its own *"files that shed em dashes are lowered or dropped from the legacy list"* assertion asks for.

### Still red, and not mine to decide

Three remain. Each needs a judgment call inside code this PR does not touch, so I am proposing rather than guessing — quietly inventing an answer in someone else's feature is worse than a red check.

**1. `typecheck:client` — 2 errors in `client/src/components/scenarios/findings/scenario-findings-derivation.ts` (lines 187, 266).** #5223 (`e5083e7`) added required `notRun` to `GoalFindingsModel` and required `launch: LaunchTotals` to the findings model, and updated the swarms producer but not the **scenarios** one, which builds the same models.

The mechanical patch is `notRun: false` and `launch: { total, succeeded: total, failed: 0, rateLimited: 0 }` — and I think that second half is wrong, which is why I stopped. That module's own header says User Testing "has real visitors who were never assigned anything" and "there is no run to key it by": the surface has no launch concept at all. `LaunchTotals` is documented as "wave-wide launch outcomes" feeding the summary and the honesty chips, so synthesising "all N sessions launched" would put a claim about launching on a surface that never launches. The honest options are to make both fields optional on the shared model, or to give the scenarios path its own model — a call for whoever owns #5223.

**2. em-dash ratchet — `components/swarms/findings/findings-headline.ts: 7, allowed 5`.** The ratchet counts em dashes in **string literals only**, so this is two new ones in user-facing copy. The fix is a copy edit in main's new feature text, and the ratchet exists precisely so that edit is made deliberately by the copy's author rather than silently by whoever's PR happens to be red. Raising the allowance would defeat the check.

**3. `SwarmsTab.createFlow` → "writes nothing until Launch, then creates personas, journeys, and one run each"** — `Unable to find an element with the text: /Running: Refund a charge/`. #5238 ("Run swarm grounding and setup in the background, with no UI") changed what the running step renders; the assertion still expects the old copy. Updating it means knowing which copy is now intended.

### Also in this push

CodeRabbit's finding on `sign-in-required.ts` was correct and is fixed — see the reply on that thread.

**Validated before pushing:** the four suites this PR owns are green (46 tests), `SwarmsTab.*` is down to that one main-side assertion, and `typecheck:client` is down to exactly the two errors above. No re-run spent: these failures are real, not flakes.

I'll keep watching and re-check as `main` moves.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcpjam-inspector/client/src/lib/sign-in-required.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  if (start !== -1 && end > start) {
+    try {
+      const parsed: unknown = JSON.parse(raw.slice(start, end + 1));
+      const fromText = messageFromRecord(parsed);
+      if (fromText) return fromText;
+    } catch {
+      // Not JSON after all — the code matched inside prose. Fall through.
+    }
+  }
+  return FALLBACK_MESSAGE;
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/lib/sign-in-required.ts]: Correct, and fixed in `128ba31`.

The over-match was real: `raw.includes(SIGN_IN_REQUIRED_CODE)` fell through to an unconditional `return FALLBACK_MESSAGE`, so any text merely containing the string classified as a refusal. The consequence is worse than a wrong message — `useRunInsights` latches `authRefused` on this result, so one misclassified error would suppress auto-requests for the rest of the session.

The code is now honoured only where a code belongs:

- a parsed payload whose `code` is the sentinel (unchanged), or
- the sentinel in a **code position** — `/["']?code["']?\s*[:=]\s*["']?sign_in_required\b/` — which is the truncated-payload case the fallback exists for.

Anything else returns `null`.

I kept the fallback rather than requiring a parse, because the shape it covers is real: `err.data` does not survive every path a rejection takes to a hook, and Convex prefixes `Server Error` onto mutation rejections. I moved the existing test's fixture from `refused: sign_in_required` to `Server Error code: sign_in_required, ...` so it pins the supported form rather than the loose one, and added the regression test you asked for — three prose cases (`Could not find public function "sign_in_required_probe"`, `TypeError: sign_in_required is not a function`, `migration sign_in_required applied`), all asserting `null`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
Update on the blockers from [my earlier comment](https://github.com/MCPJam/inspector/pull/5235#issuecomment-5708853064) — `main` moved six more commits and fixed two of the three itself. Merged in as `878ffe9`.

**Gone, fixed on `main`:**
- `typecheck:client` — the `notRun` / `launch` errors in `scenario-findings-derivation.ts`. **0 errors** on the merged tree now.
- em-dash ratchet — `findings-headline.ts` is back under its allowance.

**Still red, still `main`'s.** The ratchet now reports two different files:

```
components/shared/usage-insights/ExplanatoryFlowOptIn.tsx: 2, allowed 1
hooks/use-run-insights.ts: 2, allowed 1
```

Both are copy `main` added without lowering — or raising — its own list. The evidence is file-level rather than inferred: `"MCPJam's daily analysis budget is used up — try again after 00:00 UTC."` exists on `origin/main` and does **not** exist anywhere on this branch's pre-merge history; `ExplanatoryFlowOptIn.tsx` carries seven em dashes on `origin/main` and this PR does not touch the file.

I have not "fixed" it, for the same reason as last time: the ratchet counts em dashes in **string literals only**, so the fix is a copy edit in someone else's new feature text, and raising the allowance defeats the check the assertion exists to be. That call belongs to whoever wrote the copy.

**Also in this push — `main` superseded two of this PR's changes, and I took `main`'s:**

| Conflict | Resolution |
|---|---|
| `server/routes/web/swarm-generate.ts` | `main`'s new `upstreamRefusalRouteError` does the same forwarding across the eval adapters too, and forwards the backend's **whole** refusal envelope. This PR's `details.upstreamCode` is deleted. |
| `SwarmsTab.sessions` / `SwarmsTab.createFlow` mocks | `main` fixed its own gaps (`importOriginal` spread, `isAuthenticated: true`). The stopgaps I ported last round are dropped — they no-op now, as intended. |

The client follows the helper: `SwarmGenerateError.signInRequired` reads `details.code` instead of `details.upstreamCode`. The response's top-level `code` still cannot serve — it is the proxy's HTTP-shaped one, `FORBIDDEN` for every 403 whatever caused it. Tests on both sides updated.

**Validated before pushing:** `typecheck:client` 0 errors; **6,018 tests passing**, 1 failing — the ratchet entry above.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcpjam-inspector/client/src/lib/sign-in-required.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    try {
+      const parsed: unknown = JSON.parse(raw.slice(start, end + 1));
+      const fromText = messageFromRecord(parsed);
+      if (fromText) return fromText;
+    } catch {
+      // Not JSON after all — fall through to the code-position check, which
+      // is the only other thing that counts.
+    }
+  }
+  return CODE_IN_CODE_POSITION.test(raw) ? FALLBACK_MESSAGE : null;
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/lib/sign-in-required.ts:124]: Correct, and fixed in `e12c52a`.

Verified the exact input: `{"code":"provider_error","message":"upstream returned code: sign_in_required"}` parsed fine, `messageFromRecord` returned `null` because the code did not match — and then execution fell through to `CODE_IN_CODE_POSITION.test(raw)`, which re-read that same payload **as prose** and matched `code: sign_in_required` *inside the message string*. A sentence quoting somebody else's error classified as a refusal.

A parsed payload is now authoritative:

```ts
if (hasCode(parsed)) return messageFromRecord(parsed);
// Parsed, but not a refusal envelope — it says nothing either way, so
// the code-position scan below still gets its turn.
```

The `hasCode` gate rather than "any successful parse" is deliberate: a payload that parsed but claims no `code` at all is not a refusal envelope and should not be able to *suppress* a genuine code-position match elsewhere in the text. So the scan survives exactly where it is still the only evidence — a truncated payload — and is skipped wherever the envelope has already answered for itself.

Regression test added with your input verbatim.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcpjam-inspector/client/src/components/evals/use-insight.ts [hunk 3]
```diff
@@ -193,6 +224,7 @@ export function useInsight<TResult extends { summary?: string }>(
             }
             setUnavailable(true);
           } else {
+            setSignInRequired(classified.signInRequired === true);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/components/evals/use-insight.ts:279]: Correct, and fixed in `e12c52a`.

Confirmed the mechanism: the run-change effect does `hasAutoAttemptedRef.current = false`, and that ref is per-instance and per-run, so nothing survived to stop the next run's auto-request. A guest opening run after run sent one rejected mutation each time.

What makes this clearly right rather than a judgment call is that the sibling hook already worked the way you describe — `hooks/use-run-insights.ts` latches `authRefusedRef` on exactly this classification, and I had added the sign-in branch to that latch in this PR while leaving `useInsight` without one. The two hooks disagreed about the same refusal. They now agree:

```ts
// Refused for WHO is asking, on some earlier run. Navigating does not
// change that, so do not spend another doomed request per run opened.
if (signInRefusedRef.current) return;
```

Cleared only by an explicit press, distinguished from the auto path by the `autoClaimedRunId` argument the auto-request effect passes and a press does not — a press is the one event that can mean the viewer signed in since.

Two regression tests: a second run does not re-fire, and an explicit press does.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcpjam-inspector/client/src/components/swarms/new-swarm-create-flow.tsx [hunk 4]
```diff
@@ -1058,9 +1071,17 @@ export function NewSwarmCreateFlow({
...
+          ? err.message
+          : null;
+      setGenerateSignInRequired(signInRefusal);
+      setDescribeStepError(limitDialogRaised || signInRefusal ? null : err);
       setErrorMessage(
-        limitDialogRaised
+        limitDialogRaised || signInRefusal
           ? null
           : err instanceof SwarmTargetMaterializeError ||
               err instanceof ComposerResolveError ||
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/components/swarms/new-swarm-create-flow.tsx:1087]: Fair gap, and added in `e12c52a`.

You're right that the coverage stopped at the transport boundary: `swarm-api.test.ts` asserted the flag and message propagate, and nothing asserted what the create flow *draws* with them. Added to `SwarmsTab.createFlow.test.tsx`, next to the two existing failure tests, which are unchanged:

```ts
it("offers sign-in for a guest refusal instead of carding it", async () => {
  generateSwarmPersonaBatchMock.mockRejectedValue(
    new SwarmGenerateError(403, "Sign in to generate personas and journeys.", false, true),
  );
  …
  expect(await screen.findByText("Sign in to generate personas and journeys.")).toBeInTheDocument();
  expect(screen.getByRole("button", { name: /^Sign in$/i })).toBeInTheDocument();
  expect(screen.queryByRole("alert")).not.toBeInTheDocument();
});
```

The negative half is the part worth having: the `queryByRole("alert")` assertion is what pins "one surface for the refusal, not two saying the same thing" — the same property the existing model-limit test protects for its dialog.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
Three checks went red on `3f791da` — **Inspector Tests 3/6**, **4/6** and **Run Tests**. None of them is this PR's, and the fix for all three is carried here as `ff4f6bb`.

**What failed, and whose it is.** `main` is red on its own right now, and CI runs this branch against the merge ref, so its breakage surfaces here:

| Failure | Cause on `main` |
|---|---|
| `routes/v1/__tests__/agent.test.ts` ×2, `proposed-actions.test.ts`, `client/src/lib/__tests__/permalink-routes.test.ts` | #5240 renamed the eval surface `/evals/...` → `/evaluate/...` and left seven assertions pinning the old path |
| `services/evals/__tests__/runner-parity.test.ts` ×5 | #5246 added per-step `requestPayloadsJson`, which the golden snapshot did not carry |

This PR's diff is 14 client files plus one `server/routes/web` test. `server/routes/v1/` and `shared/` are byte-identical to `main` here (`git diff --stat origin/main HEAD -- …` is empty), so nothing on the failing path is ours.

**The fix already exists** as #5253, and is ported verbatim rather than waited on — waiting on that PR to merge is still waiting, and it no-ops the moment the base branch carries it. Two parts:

- The seven stale `/evals/suite/...` expectations, across the CLI, the SDK reporter, the client permalink registry and the two v1 route suites. The rename shipped with a legacy redirect, so what changed is the path these producers *emit*, not the contract they honour — only the expectations were left behind.
- `runner-parity`: `requestPayloadsJson` persists as a JSON **string**, so `scrub` never descends into it, and the turn ids inside embed `Date.now()`. Normalizing them pins the snapshot to the persisted shape rather than to the clock. The regenerated snapshot is additions only, and no unscrubbed `trace_turn_<digits>` survives in it.

**Verified locally on the merged tree:** `agent` + `proposed-actions` + `permalink-routes` + `runner-parity` → 131 passed; `sdk/tests/eval-run-reporter` → 35 passed; the full CLI suite → 175 passed, 0 failed (both View-link tests among them); this PR's own eight suites → 164 passed; `typecheck:client` → 0 errors.

One local-only failure is worth naming so it is not mistaken for a regression later: `services/browserd/local/__tests__/egress-proxy.test.ts` fails in this sandbox with `EAFNOSUPPORT: address family not supported ::1`. The container has no IPv6 loopback; the test is green on CI runners and the file is not in this diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcpjam-inspector/client/src/components/evals/use-insight.ts [hunk 5]
```diff
@@ -193,6 +239,8 @@ export function useInsight<TResult extends { summary?: string }>(
             }
             setUnavailable(true);
           } else {
+            if (classified.signInRequired) signInRefusedRef.current = true;
+            setSignInRequired(classified.signInRequired === true);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/components/evals/use-insight.ts]: Both races are real, and both are mine — the sticky latch introduced them. Fixed in `ba6ef71`.

Each request now captures, at request time, what it is entitled to speak for: its originating run, and the explicit-request generation current when it started.

- **Visible state** (`signInRequired`, `error`, `requested`) is gated on `runIdRef.current === originRunId`. `requested` matters as much as the banner here: run B may have a request of its own still in flight, and run A's rejection was clearing it.
- **The latch** is gated on `explicitRequestGenerationRef.current === generationAtRequest`, so a request already in flight when the viewer presses cannot set it back after the press cleared it.

One deliberate difference from the suggestion, because the three pieces answer different questions and a single guard would get one of them wrong:

- A **permanent `unavailable`** (missing backend function) is a fact about the *deployment* — not this run, not this viewer — so it is set unconditionally, however late it lands and wherever the viewer has navigated. Scoping it to the originating run would leave the panel live against a function that does not exist.
- The **latch still sets on a stale-run refusal**. The refusal was late, not wrong: a guest is still a guest on run B. Suppressing it there would restore exactly the one-doomed-request-per-run behaviour the latch exists to stop. So the guard is "no newer *explicit* request", not "same run" — and there is a test pinning that the stale refusal latches while the banner does not, so the next person cannot quietly tighten it into the earlier bug.

Three tests added. Verified they pin something: reverting the hook and re-running, `does not show run A's refusal on run B` and `an in-flight refusal cannot re-latch after an explicit press cleared it` both fail, while the third passes either way by design — it is the negative guard against over-correcting. With the fix, `use-insight` is 18/18 and the five sign-in suites are 127 passed; `typecheck:client` is clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
A second main-side breakage, fixed here as `0cbbe3a`. Not this PR's either.

`main` moved to `19085a6` (cancelling the runs on `ba6ef71` — those were **cancelled**, not failed, because a base push recomputes the merge ref) and is now red on its own **em-dash ratchet**, in three places:

| Entry | Cause on `main` |
|---|---|
| `swarm-stop-run.tsx: 1, allowed 0` | new file from #5256, carries an em dash in a toast |
| `findings-headline.ts: 4, list says 5` | shed copy without lowering its frozen count |
| `swarm-run-detail.tsx: 3, list says 4` | same |

All three files are byte-identical to `main` here (`git diff --stat origin/main HEAD --` on each is empty), and no open PR fixes it — unlike the `/evaluate` breakage above, where #5253 existed to port.

The ratchet's own header dictates the remedy, and the two cases want opposite things, which is why this isn't one uniform edit:

> Failing because you wrote a new one? Rewrite the sentence; a comma, a colon or a full stop almost always does the job. Failing because you cleaned a file up? Lower its number here, or drop the entry once it reaches zero.

So the two stale counts are lowered, and the new one is a **rewrite** — `Run stopped — N goals could not be stopped` → `Run stopped:`, a colon because the clause elaborates on the one before it. Adding `swarm-stop-run.tsx` to `LEGACY_EM_DASH_COPY` would have been the smaller diff and the wrong one: that list is a frozen snapshot of existing copy, and grandfathering a brand-new file into it retires the fence for that file permanently. The `refused.length === 0` branch is untouched, so the two suites asserting the exact string `"Run stopped"` still pass.

Also merged main's swarm reporting work (#5248), which touches three files this PR also touches. Conflict-free, and `swarm-api.ts`'s refusal path is unaffected.

Verified on the merged tree: `typecheck:client` **0 errors**; ratchet + both `SwarmsTab` suites **94 passed**; the stop-toast suites **87 passed**; this PR's own suites plus the previously ported route/snapshot fixes **298 passed**.

Worth noting for whoever picks up the `/evaluate` rename: `main` still carries the eight stale `/evals/suite/...` assertions, so #5253 is still needed there — the port in `ff4f6bb` keeps this PR readable in the meantime and no-ops once it lands.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
**Inspector Tests 1/6** failed on `0cbbe3a` with two more of `main`'s, both from #5248. Fixed here as `b749a77`. No open PR covers either, so they are ported rather than waited on.

Neither is this PR's: the whole failing path is byte-identical to `main` (nothing under `server/services/`, `server/routes/v1/__tests__/` or `docs/reference/` is in this diff), and both reproduce locally on that code.

**1. `swarm-agent.test.ts`** — the producer gained `swarm-standard-checks-v1` in its request body and the assertion did not. The expectation now follows the producer.

**2. The openapi decision-vocabulary ratchet** — this one is worth spelling out, because the tempting fix is the wrong one.

#5248 added three fresh copies of closed vocabularies to `openapi.json` — the run-level verdict reasons, the per-case ones, and the swarm report's observation stage — and registered none of them. All three have **empty descriptions**: 28 members on the wire that the only document a third-party integrator reads never defines. That is precisely the drift this ratchet was written for (its header describes `StageResultRow.reason` shipping 23 of 29 members for the same reason), so the fix is the one its own failure message prescribes, not an exemption:

- `resolveSite` learns `items`. The completeness walk already reached these sites — `items` isn't in its structure-skipping list — so a site it flagged could not be registered until the resolver agreed.
- The two verdict copies are registered separately, for the reason the friction pair already is: the run-level list and the per-case one answer different questions.
- **The descriptions are transcribed, not written.** The verdict reasons come from `EVAL_VERDICT_DECISION_REASON_LABELS` in `decision-labels.ts`; the observation stage reuses the spec's own words from `StageResultRow.stage`, which is the same vocabulary already documented. Containment doesn't apply to either vocabulary, so nothing forced this — but a spec that describes a member in different words from the SDK is the "one run, five readings" problem the ratchet's own header names, and this PR has no standing to invent a first reading for someone else's schema.

One note for whoever touches `openapi.json` next: it is edited here as **text**, three inserted lines. A `JSON.parse` → `JSON.stringify` round-trip reorders every response object keyed by HTTP status, because JS sorts integer-like keys ahead of string ones — I tried that first and got a 5,861-line diff on a published contract for three added descriptions.

Verified: shard **1/6**, the failing one, now **372 files / 5,345 tests green**; all seven other openapi guards **74 tests green**; `typecheck:client` **0 errors**.

Running total of main-side breakages carried on this PR: the `/evaluate` rename and payload snapshot (`ff4f6bb`, fix PR #5253 still unmerged), the em-dash ratchet (`0cbbe3a`), and these two. Each no-ops once `main` carries its own fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
**Run Tests** failed on `aa4de38`. Not an assertion — the CLI suite reported **1334 pass, 0 fail, 1 cancelled**, and then:

```
test at tests/tasks-commands.test.ts:1:1
✖ cli/tests/tasks-commands.test.ts (117460.748672ms)
  'Promise resolution is still pending but the event loop has already resolved'
npm run test:fast -w @mcpjam/cli exited with code SIGTERM
```

Nothing failed; the runner was killed. `tasks-commands.test.ts` took **117s** of a run that reached **298s** in total before SIGTERM, so this is a wall-clock ceiling, not a broken test.

Not this PR's: the entire CLI diff here is **two string literals** in `cli/tests/eval.test.ts` (part of the `/evaluate` rename port), `tasks-commands.test.ts` is untouched and last changed by #5124, and it passes locally on this branch — **29/29 in 73.7s**, which is the same code with a quieter machine.

I have re-run the failed jobs once, which is my first and only re-run on this PR. That re-run is the test of the hypothesis, not a shrug: 74s locally against 117s on a loaded runner says the margin is thin, so if it reddens again it is not load and the ceiling is real — at which point this is worth someone's attention regardless of this PR, because it will start taking down unrelated ones. I will say so here rather than re-running a second time.

Worth noting either way: at 298s for the CLI lane with a single file eating 117s of it, this is close enough to the edge that the next test added to `tasks-commands` may push it over on every run.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
**Correcting my last comment: `Run Tests` was mine, and it was not a wall-clock ceiling.** Fixed in `096e46d`.

The re-run failed identically (115s vs 117s), which ruled out load — and reading the full log rather than its tail showed why the CLI timing was a red herring:

```
[sdk] FAIL tests/swarm-openapi-contract.test.ts
      Error: Swarm reporting OpenAPI drift; run sdk/scripts/generate-swarm-report-schema.ts
[sdk] npm run test -w @mcpjam/sdk exited with code 1
--> Sending SIGTERM to other processes..
[cli] ✖ cli/tests/tasks-commands.test.ts (115286ms)
        'Promise resolution is still pending but the event loop has already resolved'
```

`test:parallel:rest` runs the lanes under `concurrently --kill-others-on-fail`. The **sdk** lane failed, and `concurrently` SIGTERMed the siblings; the CLI happened to be inside its slowest file when the signal landed, so it reported one cancelled test and no failures. The CLI was collateral. `tasks-commands.test.ts` is fine, there is no ceiling, and my note about the 298s lane was reading a symptom. Apologies for the noise — the lesson worth keeping is that a SIGTERMed lane with `fail 0` means you want the log of whichever lane exited 1, not the one that got killed.

**The actual defect was mine, from `b749a77`.** `SwarmReport` and `EvalVerdictDecision` are **generated** into `openapi.json` by `sdk/scripts/generate-swarm-report-schema.ts`, and I added those three vocabulary descriptions by editing the spec by hand. The generator's output stopped matching the file, which is exactly what `swarm-openapi-contract` exists to catch. It was the wrong edit even where it was green: a description hand-written into a generated region is erased by the next run of the generator.

They now live in the generator, which already sets a description this way for `JourneyRunSession.outcome`. The words are unchanged — the reasons still come from `EVAL_VERDICT_DECISION_REASON_LABELS`, and the stage now *copies* `StageResultRow.stage`'s description off the document instead of duplicating the text, so the two sites for one vocabulary cannot drift apart.

The labels are read in the script rather than via `.describe()` on the Zod schema. That was my first attempt and it is a cycle: `decision-labels.ts` already imports from `verdict-policy.ts`.

Verified: generator `--check` clean; all **7** openapi guards **99 tests** green; the three SDK contract suites **66 tests** green; full SDK suite **8,722 passed**. The one remaining SDK failure there, `browser-entry-guard`, is not this change — it fails identically with my edits stashed and it passed in CI's own sdk lane on the previous run.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by chelojimenez** [comment]
Merged main at `8b64ee7`. **Every ported fix this PR was carrying has now dropped out**, which is what a port is supposed to do.

`main` landed its own versions: #5253 as `def5fb6`, and `8ca1417` for the rest. Four conflicts, main's side kept in all four:

- `swarm-agent.test.ts` and `openapi-decision-vocabularies.test.ts` — main reached the *same* fix independently, down to teaching `resolveSite` to walk `items` and registering the three sites. Only the comments and a branch order differ.
- `swarm-runner.rubric.test.ts` and `openapi.json` — same.

The generator change went too, and that one is worth a note. Main put the three vocabulary descriptions on the Zod schemas with `.describe()` — the approach I tried first and abandoned, because importing `EVAL_VERDICT_DECISION_REASON_LABELS` into `verdict-policy.ts` is a cycle. Main avoided the cycle by **inlining** the words rather than importing them, which is the trade I chose not to make. That makes my generator-side copy redundant, and keeping both would have been drift with extra steps — two producers writing the same three descriptions, disagreeing by ten characters. `generate-swarm-report-schema.ts` is restored to main's.

**The PR's diff against main is now exactly its own 16 files** — the five sign-in surfaces, their tests, and the swarm-generate route test. Nothing borrowed left in it.

Validated on the merged tree: generator `--check` clean; the six formerly conflicting suites **162 tests** green; this PR's nine suites plus the em-dash ratchet **169 tests** green; `swarm-openapi-contract` green; `typecheck:client` **0 errors**.

---
_Generated by [Claude Code](https://claude.ai/code)_
